### PR TITLE
Enable splitting into train, validation, and test

### PR DESCRIPTION
**Pull request by smcolby**
## Description
For deep learning models, optimal training requires checkpointing against validation set performance. This PR will allow for three-way splits into train, validation, and test sets. 

While not strictly backwards compatible with original two-way splits, requesting train + validation only, or train + test only, is valid usage -- "empty" splits will simply be returned as `None`, and will be handled appropriately by the relevant trainer. Empty validation _and_ test splits will raise a value error (i.e. both cannot be zero).

## Todos
Notable points that this PR has either accomplished or will accomplish.
  - [X] Implement three-way split logic
  - [ ] Add additional checkpoint logic
  - [ ] Update unit tests
  - [ ] Ensure functional end-to-end training run (e.g. multitask mtenn)

## Status
- [ ] Ready to go

### CODE REVIEW

**Comment by smcolby** [comment]
Resolves #78

File: openadmet/models/anvil/workflow.py [hunk 1]
```diff
@@ -217,39 +302,66 @@ def run(self, output_dir: PathLike = "anvil_run", debug: bool = False) -> Any: .
 
 
 class AnvilWorkflow(AnvilWorkflowBase):
+    """
+    Workflow for running basic Anvil configuration.
+    """
+
     driver: Drivers = Drivers.SKLEARN
 
+    @model_validator(mode="after")
```
Comment (hmacdope) [inline openadmet/models/anvil/workflow.py:311]: Nice

File: openadmet/models/anvil/workflow.py [hunk 2]
```diff
@@ -453,8 +633,12 @@ def run(
                 y_train=train_dataloader,
                 use_wandb=use_wandb,
                 tag=model_tag,
+                target_labels=target_labels,
             )
+
+            # Write evaluation report
             eval.report(write=True, output_dir=output_dir)
+
         logger.info("Evaluation done")
```
Comment (hmacdope) [inline openadmet/models/anvil/workflow.py:643]: These all look really good, thanks for commenting everything also

File: openadmet/models/eval/classification.py [hunk 3]
```diff
@@ -74,7 +74,19 @@ def evaluate(self, y_true=None, y_pred=None, use_wandb=False, tag=None, **kwargs
                     _y_pred = y_pred[:, 1].ravel()
                     _y_true = y_true.ravel()
 
-            # Multiclass
+            # Also binary case
+            elif y_true.ndim == 2 and y_true.shape[1] == 1:
```
Comment (hmacdope) [inline openadmet/models/eval/classification.py:78]: Can this not be folded into the branch above? I know its a bit crowded but DRY would say combined them. 

I could be misinterpreting.

File: openadmet/models/eval/classification.py [hunk 4]
```diff
@@ -204,6 +216,10 @@ def roc_curve(
         if y_true.ndim == 1:
             fpr, tpr, _ = roc_curve(y_true.ravel(), y_pred[:, 1].ravel())
 
+        # Also binary case
```
Comment (hmacdope) [inline openadmet/models/eval/classification.py:219]: Same here, can we fold into above branch?

File: openadmet/models/eval/classification.py [hunk 5]
```diff
@@ -234,6 +250,12 @@ def pr_curve(
                 y_true.ravel(), y_pred[:, 1].ravel()
             )
 
+        # Also binary case
```
Comment (hmacdope) [inline openadmet/models/eval/classification.py:253]: Same here

File: openadmet/models/eval/regression.py [hunk 6]
```diff
@@ -33,53 +34,68 @@ class RegressionMetrics(EvalBase):
...
-    def evaluate(self, y_true=None, y_pred=None, use_wandb=False, tag=None, **kwargs):
+    def evaluate(self, y_true=None, y_pred=None, use_wandb=False, tag=None, target_labels=None, **kwargs):
         """
         Evaluate the regression model
         """
         if y_true is None or y_pred is None:
             raise ValueError("Must provide y_true and y_pred")
 
+        n_tasks = y_true.shape[1]
+        assert n_tasks == y_pred.shape[1]
```
Comment (hmacdope) [inline openadmet/models/eval/regression.py:45]: Raise a more descriptive error here.

File: openadmet/models/eval/regression.py [hunk 7]
```diff
@@ -123,13 +146,20 @@ def make_stat_caption(self):
...
-        for metric in self.metric_names:
-            value = self.data[metric]["value"]
-            lower_ci = self.data[metric]["lower_ci"]
-            upper_ci = self.data[metric]["upper_ci"]
-            confidence_level = self.data[metric]["confidence_level"]
-            stat_caption += f"{self._metrics[metric][2]}: {value:.2f}$_{{{lower_ci:.2f}}}^{{{upper_ci:.2f}}}$\n"
-        stat_caption += f"Confidence level: {confidence_level}"
+
+        for task_name in self.task_names:
+            if task_name == 'tag':
```
Comment (hmacdope) [inline openadmet/models/eval/regression.py:151]: What is the role of this?

File: openadmet/models/eval/regression.py [hunk 8]
```diff
@@ -247,6 +294,22 @@ def regplot(
 
         return fig
 
+    def make_stat_caption_from_data(self, data):
```
Comment (hmacdope) [inline openadmet/models/eval/regression.py:297]: This is not used AFAIK? Could be wrong, should be simple to fold in.

File: openadmet/models/features/chemprop.py [hunk 9]
```diff
@@ -27,8 +27,9 @@ def featurize(self, smiles: Iterable[str], y: Iterable[Any]=None) -> DataLoader:
         """
         Featurize a list of SMILES strings
         """
-        if y:
-            y = y.to_numpy().reshape(-1, 1)
+        if y is not None:
+            #fix broken for multi-task
```
Comment (hmacdope) [inline openadmet/models/features/chemprop.py:31]: No longer true?

File: openadmet/models/split/scaffold.py [hunk 10]
```diff
@@ -9,45 +9,179 @@ class ScaffoldSplitter(SplitterBase):
...
+            return (
+                X[train_idx],
+                X[val_test_idx],
+                None,
+                y[train_idx],
+                y[val_test_idx],
+                None,
+            )
+
+        # If both test and validation sets are requested, split the remaining data
```
Comment (hmacdope) [inline openadmet/models/split/scaffold.py:48]: Nitpick but would prefer this had a guard for its condition, i.e `if self.test_size and self.val_size`

File: openadmet/models/split/scaffold.py [hunk 11]
```diff
@@ -9,45 +9,179 @@ class ScaffoldSplitter(SplitterBase):
...
```
Comment (hmacdope) [inline openadmet/models/split/scaffold.py:110]: Same here, not sure if clustering twice is how this is supposed to work, but I think impact should be minimal

Comment (hmacdope) [reply openadmet/models/split/scaffold.py:110]: Lets raise as an issue.

File: openadmet/models/split/scaffold.py [hunk 12]
```diff
@@ -9,45 +9,179 @@ class ScaffoldSplitter(SplitterBase):
...
```
Comment (hmacdope) [inline openadmet/models/split/scaffold.py:171]: Same here with guard

File: openadmet/models/split/sklearn.py [hunk 13]
```diff
@@ -9,15 +9,35 @@ class ShuffleSplitter(SplitterBase):
...
+
+        # If no validation set is requested, return train and test sets
+        if self.val_size == 0:
+            return X_train, None, X_val_test, y_train, None, y_val_test
+
+        # If no test set is requested, return train and validation sets
+        if self.test_size == 0:
+            return X_train, X_val_test, None, y_train, y_val_test, None
+
+        # If both test and validation sets are requested, split the remaining data
```
Comment (hmacdope) [inline openadmet/models/split/sklearn.py:34]: Same here

**Review by hmacdope — CHANGES_REQUESTED** [review]
Looking great, just some small changes. :)

File: openadmet/models/split/scaffold.py (hunk 11 above)
Comment (hmacdope) [reply openadmet/models/split/scaffold.py:110]: Does it instead make sense to just split the test ratio i.e split into (test, train) and then test becomes -> (test, val)

File: openadmet/models/eval/classification.py (hunk 3 above)
Comment (hmacdope) [reply openadmet/models/eval/classification.py]: I have made this change @smcolby, feel free to revert

File: openadmet/models/eval/regression.py (hunk 6 above)
Comment (hmacdope) [reply openadmet/models/eval/regression.py]: I have made this change @smcolby

File: openadmet/models/features/chemprop.py (hunk 9 above)
Comment (hmacdope) [reply openadmet/models/features/chemprop.py]: Fixed

File: openadmet/models/split/scaffold.py (hunk 10 above)
Comment (hmacdope) [reply openadmet/models/split/scaffold.py:48]: Can ignore this if you disagree

File: openadmet/models/eval/classification.py (hunk 3 above)
Comment (smcolby) [reply openadmet/models/eval/classification.py]: I had this as two if statements because I thought that checking `y_true.shape[1]` would cause an error if `ndim` = 1 (i.e. index out of range), but looks like this usage is fine after a quick check.

File: openadmet/models/eval/classification.py (hunk 4 above)
Comment (smcolby) [reply openadmet/models/eval/classification.py:207]: Propagating above change here.

File: openadmet/models/eval/regression.py (hunk 7 above)
Comment (smcolby) [reply openadmet/models/eval/regression.py:152]: I'm not sure, didn't write this. Looks like it just skips evaluation if the task key is `tag`.

File: openadmet/models/eval/regression.py (hunk 8 above)
Comment (smcolby) [reply openadmet/models/eval/regression.py:299]: Also did not author. Do you suggest removal until we'd make use of it?

File: openadmet/models/split/scaffold.py (hunk 11 above)
Comment (smcolby) [reply openadmet/models/split/scaffold.py:110]: I think we should chat briefly on this, not sure I 100% follow

File: openadmet/models/split/scaffold.py (hunk 10 above)
Comment (smcolby) [reply openadmet/models/split/scaffold.py:48]: The sum of `test_size` and `val_size` cannot be zero, which is checked at construction in `SplitterBase`. So this branch of the if tree is implicitly guarded for. I made the if:elif:else logic explicit, but did not add additional checks for `test_size` and `val_size`. Sound good?

Comment (hmacdope) [reply openadmet/models/split/scaffold.py:48]: Yep

File: openadmet/models/eval/regression.py (hunk 7 above)
Comment (hmacdope) [reply openadmet/models/eval/regression.py:152]: Issue raised #171, lets deal with it later.

File: openadmet/models/split/scaffold.py (hunk 11 above)
Comment (hmacdope) [reply openadmet/models/split/scaffold.py:111]: Lets push to an issue and discuss later this week.
